src/otk/traversal.py

```diff
@@ -1,24 +1,52 @@
+import copy
+import inspect
+import os
+import pathlib
+
+from .error import CircularIncludeError
+
+
 class State:
-    def __init__(self, path, define_subkeys, includes=None):
+
+    def __init__(self, path: pathlib.Path = None):
+        if path is None:
+            path = pathlib.Path()
         self.path = path
-        self.define_subkeys = define_subkeys
-        if includes is None:
-            includes = []
-        self.includes = includes
+        self._define_subkeys = []
+        self._includes = []
+        if path != pathlib.Path():
+            self._includes.append(path)
 
-    def copy(self, *, path=None, define_subkeys=None, includes=None) -> "State":
+    def copy(self, *, path=None, subkey_add: str = None) -> "State":
         """
-        Return a new State, optionally redefining the path and define_subkeys
-        properties. Properties not defined in the args are (shallow) copied
+        Return a new State, optionally redefining the path and add a define
+        subkey. Properties not defined in the args are (shallow) copied
         from the existing instance.
-        The 'define_subkeys' keeps track how deep in a define the recursion
-        is and can be used to modify the global context.
         """
-        if path is None:
-            path = self.path
-        if define_subkeys is None:
-            define_subkeys = self.define_subkeys
-        if includes is None:
-            includes = self.includes.copy()
+        new_state = copy.deepcopy(self)
+        if subkey_add:
+            new_state._define_subkeys.append(subkey_add)
+        if path:
+            if path in self._includes:
+                circle = [os.fspath(p) for p in new_state._includes]
+                raise CircularIncludeError(f"circular include from {circle}")
```

Comment:
Nitpick: would this be nicer as:
```python
f"circular include: ' -> '.join(circle)
```
and also if `circle.append(path)` so we get:
```
circular include: a.yaml -> b.yaml -> c/c.yaml -> a.yaml
```

Reply:
Let's merge and do this in another PR if we like it.